private/bufpkg/bufcheck/buflint/internal/buflintvalidate/numeric.go

```diff
@@ -0,0 +1,274 @@
+		)
+		return nil
+	}
+	if compareFunc(upperBound, lowerBound) <= 0 {
+		adder.addForPathsf(
+			[][]int32{
+				{ruleFieldNumber, lowerBoundFieldNumber},
+				{ruleFieldNumber, upperBoundFieldNumber},
+			},
+			"Field %q has a %s (%v) higher than its %s (%v).",
+			adder.fieldName(),
+			adder.getFieldRuleName(ruleFieldNumber, lowerBoundFieldNumber),
+			formatFunc(lowerBound),
+			adder.getFieldRuleName(ruleFieldNumber, upperBoundFieldNumber),
+			formatFunc(upperBound),
+		)
+	}
+	return nil
+}
+
+func getNumericPointerFromValue[
+	T int32 | int64 | uint32 | uint64 | float32 | float64,
+](value protoreflect.Value) (*T, string, error) {
+	number, ok := value.Interface().(T)
+	if !ok {
+		return nil, "", fmt.Errorf("unable to cast value to type %T", number)
+	}
+	return &number, "", nil
+}
+
+func getTimestampFromValue(value protoreflect.Value) (*timestamppb.Timestamp, string, error) {
+	bytes, err := proto.Marshal(value.Message().Interface())
+	if err != nil {
+		return nil, "", err
+	}
+	timestamp := &timestamppb.Timestamp{}
+	err = proto.Unmarshal(bytes, timestamp)
+	if err != nil {
+		return nil, "", err
+	}
+	timestampErr := timestamp.CheckValid()
+	if timestampErr == nil {
+		return timestamp, "", nil
+	}
+	return nil, timestampErr.Error(), nil
+}
+
+func getDurationFromValue(value protoreflect.Value) (*durationpb.Duration, string, error) {
+	bytes, err := proto.Marshal(value.Message().Interface())
+	if err != nil {
+		return nil, "", err
+	}
+	duration := &durationpb.Duration{}
+	err = proto.Unmarshal(bytes, duration)
+	if err != nil {
+		return nil, "", err
+	}
+	if durationErrString := checkDuration(duration); durationErrString != "" {
+		return nil, durationErrString, nil
+	}
+	return duration, "", nil
+}
+
+func compareNumber[T int32 | int64 | uint32 | uint64 | float32 | float64](a *T, b *T) float64 {
+	return float64(*a - *b)
+}
+
+func compareTimestamp(t1 *timestamppb.Timestamp, t2 *timestamppb.Timestamp) float64 {
+	if t1.Seconds > t2.Seconds {
+		return 1
+	}
+	if t1.Seconds < t2.Seconds {
+		return -1
+	}
+	return float64(t1.Nanos - t2.Nanos)
+}
+
+func compareDuration(d1 *durationpb.Duration, d2 *durationpb.Duration) float64 {
+	if d1.Seconds > d2.Seconds {
+		return 1
+	}
+	if d1.Seconds < d2.Seconds {
+		return -1
+	}
+	return float64(d1.Nanos - d2.Nanos)
+}
+
+func checkDuration(duration *durationpb.Duration) string {
+	// This is slightly smaller than MaxInt64, 9,223,372,036,854,775,807,
+	// but 9,223,372,036,854,775,428 is the maximum value that does not cause a
+	// runtime error in protovalidate.
+	maxDuration := &durationpb.Duration{
+		Seconds: 9223372036,
+		Nanos:   854775428,
+	}
+	minDuration := &durationpb.Duration{
+		Seconds: -9223372036,
+		Nanos:   -854775428,
+	}
+	secs := duration.GetSeconds()
+	nanos := duration.GetNanos()
+	switch {
+	case nanos <= -1e9 || nanos >= +1e9:
+		return fmt.Sprintf("duration (%v) has out-of-range nanos", duration)
```

Comment:
`duration %v must be in the range %v to %v`, duration, min, max

Reply:
resolved in [6a37deb](https://github.com/bufbuild/buf/pull/2505/commits/6a37debb7800c2b4cd1dee0520f712038648a63a)